src/sentry/release_health/metrics_sessions_v2.py

```diff
@@ -312,15 +311,72 @@ def get_values(
             ]
 
 
+class _LimitState:
+    """Keeps track of how to limit each query
+    Before the first query, limiting_conditions = None, so the first query
+    will apply an ORDER BY ... LIMIT.
+    After that, the groups returned by the first query will be added as a
+    WHERE clause to any secondary query.
+    """
+
+    def __init__(self) -> None:
+        self._initialized = False
+        self.skip_columns: List[Column] = []  # Will not be added to limiting conditions
+
+    def update(self, groupby: Collection[Column], snuba_rows: _SnubaData) -> None:
+        if self._initialized:
+            return
+
+        # Only "totals" queries may set the limiting conditions:
+        assert Column(TS_COL_GROUP) not in groupby
+
+        def sort_key(col: Column) -> str:
+            return col.name  # type: ignore. But why?
+
+        groupby = sorted(groupby, key=sort_key)
+
+        self._groupby = sorted(groupby, key=sort_key)
+        self._groups = [{column.name: row[column.name] for column in groupby} for row in snuba_rows]
```

Comment:
Are there some references you are concerned with mutating here or should this use `self._groupby` instead?

Reply:
Also what's the point of sorting here?

Reply:
I wanted to make sure that order is preserved, so you don't end up with conditions like `(release, environment) IN [(environment1, release1)]`. But sorting is overkill -- converted to list instead.